### ISSUE TITLE
Try to include a GPT system 

### ISSUE THREAD

**Issue by alonsovidales** [DEAN-ngo/elimupi#106]
When there is no internet available, the next best thing could be a GPT model where the people could ask and get information in an easy way from multiple sources but without having to download these.
These systems will very likely became part of the day to day in any work in the near future so it could be also interesting to learn how to interact and use them in an early age.

Recently Meta released the weights for the LlaMa models: https://arxiv.org/pdf/2302.13971.pdf These models contain huge amounts of information condensed in just few GBs, in this case the model was trained with:
```
Dataset Sampling prop. Epochs Disk size
CommonCrawl 67.0% 1.10 3.3 TB
C4 15.0% 1.06 783 GB
Github 4.5% 0.64 328 GB
Wikipedia 4.5% 2.45 83 GB
Books 4.5% 2.23 85 GB
ArXiv 2.5% 1.06 92 GB
StackExchange 2.0% 1.03 78 GB
```
We should be able to run the 7GB model in a Raspberry Pi 4 ( https://twitter.com/miolini/status/1634982361757790209 ) using the Alpaca project: https://medium.com/@martin-thissen/llama-alpaca-chatgpt-on-your-local-computer-tutorial-17adda704c23
We could also consider an Orange Pi as solution for bigger models, it is more expensive but it has double the cores and memory. Also because it seems to go a bit slow in the Raspberry Pi, but I may be able to speed it up using the GPU, Alpaca only runs on CPU.

This would be like, instead of downloading all that content, have someone to ask that have read and memorised it, you can ask for book recommendations based in your preferences, solve real world issues like how to fix a car, plumbing and so on, ask historic events, explain subjects like mathematics, physics, biology..., write letters and many other applications.

My main concern is that this is an unrestricted model trained with stuff from internet. This means that it can contain missinformation and biases, it is also prone to `hallucinate incorrect answers` so this could be a bit controversial. I guess that someone will release any kind of control tool as ChatGPT has to prevent some content to be disclosed and ban problematic prompts, but by the moment I don't think there is anything, but I could ask in Meta and see if they have any kind of solution for this problem in mind.

**Comment by alonsovidales** [comment]
This is just to try to make it work and see how it goes first :)

**Comment by alonsovidales** [comment]
After trying different models, the best candidate is Alpaca, this is a model based in the 7B LlaMa model developed by Standford: https://github.com/tatsu-lab/stanford_alpaca

It is super lightweight and I managed to make it run in ElimuPi with very little trouble. So far it is able to answer everything I ask with very accurate answers:

<img width="1440" alt="Screen Shot 2023-03-25 at 23 49 25" src="https://user-images.githubusercontent.com/1468785/227745978-8259af5c-73a2-4e82-81d4-436c6f94e957.png">

The size of the model is 4G only, but it requires about 8GB of RAM to be able to run, so in the Raspberry Pi with 4GB it starts swapping what makes it extremely slow. I will try to get an Orange Pi with 16GB to test it but the price of an Orange Pi 16GB is 150 euros... I will get it anyways just to play a bit.

**Comment by alonsovidales** [comment]
I got the Orange Pi today, it works great and it is able to run all the content we have, the only problem is that it doesn't have WiFi, but it even has a M.2 socket that would be way faster and reliable than using a SD + USB disk.

It is able to load and run the 7B Alpaca model, this is a video of it running in real time:
 - https://youtu.be/ETV_oBPxncs
In the video I run the model using only 3 of the 8 cores to see if it would be possible to run it in a Raspberry Pi 4 8GB (I only have the 4GB) and leaving one of the cores for other tasks.

It takes about 20 seconds to answer and it generates about two tokens per second, this can be improved a bit removing context and so on.
It is usable but I don't think that allowing more than one execution in parallel is a good idea.

The model as it is really good after being tuned by Standford, it is very accurate and I tried to make it say nasty stuff but it is very polite as well, we could add a big disclaimer anyways specifying that it may hallucinate or spread misinformation since it was trained with public data from Internet.
We can also add a Web UI, the only one I found a bit good is: https://github.com/ViperX7/Alpaca-Turbo/ but it has too many options, we can remove everything and leave only a simple chat.

I think that we could add it as part of the distribution checking if the device has at least 8GB of memory, it is a very useful tool.

**Comment by alonsovidales** [comment]
Now I tried with Llama 3, the new model by Meta that outperforms GPT 3.5 and it works great in a Raspberry Pi 5, this is a little video: https://www.youtube.com/watch?v=ywLUZH_uNE0

As it can be seeing in the video, it is very accurate in the answers still making up some stuff when I ask about book recommendations, some of the books doesn't exist or the name of the authors is not correct. But by the rest it seems very accurate and it can translate, create stories and explain many different concepts.
It is also properly aligned, so if you ask dangerous stuff like in the video, how to make napalm, it refuses to answer.
This could be a very good tool for teachers to prepare lessons or kids with curiosity to learn about anything. 

The only thing is that it is necessary to add a heat sink with a fan to the Raspberry Pi since if not it overheats, with the heat sink I installed it works without any issue, but I'm not sure how it will go in Africa, perhaps get the best heat sink available (they are more or less cheap, like 20 euros)

The memory consumption is around 750MB and the model is a bit more of 4GB, so it could be included with the current distribution, but when the model is running, the CPU goes to 100%, so it will make unusable for other proposals.

In the video I use the terminal, but we could add a web interface, there are many for Ollama, so it should be easy.
